moveit_ros/moveit_servo/src/pose_tracking.cpp

```diff
@@ -220,59 +216,67 @@ bool PoseTracking::haveRecentEndEffectorPose(const double timespan)
 
 bool PoseTracking::satisfiesPoseTolerance(const Eigen::Vector3d& positional_tolerance, const double angular_tolerance)
 {
+  std::lock_guard<std::mutex> lock(target_pose_mtx_);
   double x_error = target_pose_.pose.position.x - command_frame_transform_.translation()(0);
   double y_error = target_pose_.pose.position.y - command_frame_transform_.translation()(1);
   double z_error = target_pose_.pose.position.z - command_frame_transform_.translation()(2);
 
-  return (fabs(x_error) < positional_tolerance(0)) && (fabs(y_error) < positional_tolerance(1)) &&
-         (fabs(z_error) < positional_tolerance(2) && fabs(angular_error_) < angular_tolerance);
+  return ((std::abs(x_error) < positional_tolerance(0)) && (std::abs(y_error) < positional_tolerance(1)) &&
+         (std::abs(z_error) < positional_tolerance(2)) && (std::abs(angular_error_) < angular_tolerance));
 }
 
 void PoseTracking::targetPoseCallback(const geometry_msgs::PoseStampedConstPtr& msg)
 {
-  // Transform to planning frame
+  std::lock_guard<std::mutex> lock(target_pose_mtx_);
   target_pose_ = *msg;
-  geometry_msgs::TransformStamped target_to_planning_frame;
+
+  // If the target pose is not defined in planning frame, transform the target pose.
   if (target_pose_.header.frame_id != planning_frame_)
   {
     try
     {
-      target_to_planning_frame = transform_buffer_.lookupTransform(planning_frame_, target_pose_.header.frame_id,
-                                                                   ros::Time(0), ros::Duration(0.1));
+      geometry_msgs::TransformStamped target_to_planning_frame = transform_buffer_.lookupTransform(
+          planning_frame_, target_pose_.header.frame_id, ros::Time(0), ros::Duration(0.1));
```

Comment:
I would actually argue that instead of `ros::Time(0)`, the transform timestamp should match the timestamp of the latest pose message. The reason being that if the target frame and planning frame move quickly w.r.t each other, the difference between latest transform and the transform at the time target pose was sent might be noticeable.

Reply:
I would argue that's exactly the point why we should take the current time. Otherwise, we would add a lag when tracking a moving target.

Reply:
Well, the most recent available pose is going to be tracked regardless.

I'm inclined to say we should use target_pose_.header.stamp rather than `ros::Time(0)`, so we check if the target pose was _sent_ recently. In other words, account for network latency.

But, I'm ok either way.

Reply:
@henningkayser you are right. If the target pose frame and the planning frame happen to be different yet static w.r.t. each other, then it makes no difference. If the frames are different and _are_ moving, to eliminate the effect of the network latency and possible other delays that happen before call to `publish`, latest possible transform should be used. Now that I slept on it, I'm not sure what made me think we should use old timestamps.